### ISSUE TITLE
`unscope` causes too many parameters to be supplied when used with subquery

### ISSUE THREAD

**Issue by Midnighter** [rails/rails#26323]
### Steps to reproduce

I've created [this minimal example](https://gist.github.com/Midnighter/8ddfad43ea22515f8a4bf3d480d6d58e).
### Expected behavior

In the test case `test_subquery_unscope`  `p2` and `p3` are `ActiveRecord::Relation`s thus causing a subquery with a join to be used. Using `unscope` to remove the first condition, I expect the query to work and only the parameters from `p3` to be applied.
### Actual behavior

The SQL generated includes both the parameters from `p2` _and_ `p3` thus supplying too many parameters and causing an `ActiveRecord::StatementInvalid` error. As can be seen in `test_avoid_subquery_unscope`, simplifying the first condition by using `p2.first` and thus avoiding a subquery there, unscope works as expected and the new condition with a subquery can be applied.
### System configuration

**Rails version**: master but also on 4.2.6 (haven't tested any others)

**Ruby version**:  ruby 2.3.1p112 (2016-04-26) [x86_64-linux-gnu]

**Comment by sgrif** [comment]
Thank you for the report. This has been fixed in Rails 5. Unfortunately the fix cannot be safely backported to 4.2.

**Comment by Midnighter** [comment]
I  still get this error
```
Using rails 5.1.0.alpha from git://github.com/rails/rails.git (at master@dd65843)
```
is the fix on a different branch?
